packages/kbn-optimizer/src/optimizer.ts

```diff
@@ -0,0 +1,294 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { inspect } from 'util';
+
+import * as Rx from 'rxjs';
+import { map, mergeMap, share } from 'rxjs/operators';
+
+import { observeWorker, WorkerStatus } from './observe_worker';
+import { OptimizerConfig } from './optimizer_config';
+import { getOptimizerVersion } from './get_optimizer_version';
+import { CompilerMsg, WorkerMsg, pipeClosure, Bundle, maybeMap } from './common';
+import { assignBundlesToWorkers } from './assign_bundles_to_workers';
+import { Watcher, ChangesStarted, Changes } from './watcher';
+import { getMtimes } from './get_mtimes';
+
+export interface OptimizerMsg {
+  state: OptimizerState;
+  event?: ChangesStarted | Changes | WorkerMsg | WorkerStatus;
+}
+
+export interface OptimizerState {
+  phase: 'initialized' | 'reallocating' | 'success' | 'running' | 'issue';
+  version: string;
+  startTime: number;
+  durSec: number;
+  compilerStates: CompilerMsg[];
+  onlineBundles: Bundle[];
+  offlineBundles: Bundle[];
+}
+
+export interface OptimizerStartedWorker {
+  type: 'worker started';
+  bundles: Bundle[];
+}
+
+export interface OptimizerComingOnline {
+  type: 'bringing bundle online';
+  bundle: Bundle;
+}
+
+export { ChangesStarted };
+export type OptimizerNotif = OptimizerStartedWorker;
+
+const msToSec = (ms: number) => Math.round(ms / 100) / 10;
+
+const getPhase = (states: CompilerMsg[]): OptimizerState['phase'] => {
+  const types = states.map(s => s.type);
+
+  if (types.includes('running')) {
+    return 'running';
+  }
+
+  if (types.includes('compiler issue')) {
+    return 'issue';
+  }
+
+  if (types.every(s => s === 'compiler success')) {
+    return 'success';
+  }
+
+  throw new Error(`unable to summarize bundle states: ${JSON.stringify(states)}`);
+};
+
+export class Optimizer {
+  constructor(private readonly config: OptimizerConfig) {}
+
+  async getCachedBundles(optimizerVersion: string) {
+    const eligible = this.config.bundles.filter(
+      // only get the mtimes for files if the bundle was built
+      // with our version of the optimizer and there is a cache key
+      b => b.cache.getOptimizerVersion() === optimizerVersion && !!b.cache.getKey()
+    );
+
+    if (!eligible.length) {
+      return [];
+    }
+
+    const mtimes = await getMtimes(
+      new Set<string>(
+        eligible.reduce(
+          (acc: string[], bundle) => [...acc, ...(bundle.cache.getReferencedFiles() || [])],
+          []
+        )
+      )
+    );
+
+    return eligible.filter(
+      bundle =>
+        bundle.cache.getKey() ===
+        bundle.createCacheKey(bundle.cache.getReferencedFiles() || [], mtimes)
```

Comment:
We should really have some way of detecting why the cache was not valid, maybe we should store the cache keys as plain text and render a diff of some sort to the log when they're different.